src/main/java/frc/trigon/robot/utilities/ShootingCalculations.java

```diff
@@ -0,0 +1,103 @@
+package frc.trigon.robot.utilities;
+
+import edu.wpi.first.math.geometry.Translation2d;
+
+import java.util.ArrayList;
+import java.util.Objects;
+
+/**
+ * This class is used to calculate the shooting angle and distance for the shooter.
+ */
+public class ShootingCalculations {
+    private static final ArrayList<Waypoint> waypoints = new ArrayList<>();
+
+    /**
+     * @param distance the distance from the target
+     * @return a waypoint with the distance, velocity and angle from the given distance
+     */
+    public static Waypoint getWaypointFromDistance(double distance) {
+        Waypoint[] waypointsFromDistance = getWaypointsFromDistance(distance);
+        if(waypointsFromDistance == null) {
+            return null;
+        }
+        double velocity = calculateBetweenTranslations(
+                new Translation2d(waypointsFromDistance[0].distance, waypointsFromDistance[0].velocity),
+                new Translation2d(waypointsFromDistance[1].distance, waypointsFromDistance[1].velocity),
+                distance
+        );
+        double angle = calculateBetweenTranslations(
+                new Translation2d(waypointsFromDistance[0].distance, waypointsFromDistance[0].angle),
+                new Translation2d(waypointsFromDistance[1].distance, waypointsFromDistance[1].angle),
+                distance
+        );
+        return new Waypoint(distance, velocity, angle);
+    }
+
+    /**
+     * Gets the velocity needed for the shooter to shoot from a distance.
+     *
+     * @param distance the distance from the target
+     * @return the velocity needed for the shooter
+     */
+    public static double getShootingVelocityFromDistance(double distance) {
+        return Objects.requireNonNullElse(getWaypointFromDistance(distance), new Waypoint(0, 0, 0)).velocity;
+    }
+
+    /**
+     * Gets the angle needed for the shooter to shoot from a distance.
+     *
+     * @param distance The distance from the target
+     * @return the angle needed for the shooter
+     */
+    public static double getShootingAngleFromDistance(double distance) {
+        return Objects.requireNonNullElse(getWaypointFromDistance(distance), new Waypoint(0, 0, 0)).angle;
```

Comment:
I don't want it to return 0, I want it to continue off the line of the two nearest waypoints, even if they're on one side of the distance.

Reply:
This will only return 0 if there's no given waypoints from getNearstWaypointsFromDistance and there are no near waypoints to null

Reply:
When will `getNearstWaypointsFromDistance` return null?

Reply:
Oh I understand